src/Workspaces/Core/Portable/Shared/Extensions/ICodeDefinitionFactoryExtensions.cs

```diff
@@ -102,21 +104,54 @@ internal static partial class ICodeDefinitionFactoryExtensions
                 // Special case.  If we're generating a struct constructor, then we'll need
                 // to initialize all fields in the struct, not just the ones we're creating.  To
                 // do that, we call the default constructor.
-                var realFields = containingTypeOpt.GetMembers()
-                                     .OfType<IFieldSymbol>()
-                                     .Where(f => !f.IsStatic);
-                var initializedFields = parameterToExistingFieldMap.Values
-                                            .OfType<IFieldSymbol>()
-                                            .Where(f => !f.IsImplicitlyDeclared && !f.IsStatic);
-                if (initializedFields.Count() < realFields.Count())
+
+                var numFieldsRequiringInitialization = 0;
+
+                var autoProperties = new List<IPropertySymbol>();
+                foreach (var containingTypeField in containingTypeOpt.GetMembers().OfType<IFieldSymbol>())
+                {
+                    if (!containingTypeField.IsStatic)
+                    {
+                        numFieldsRequiringInitialization++;
+
+                        if (containingTypeField.AssociatedSymbol is IPropertySymbol autoProperty)
+                        {
+                            autoProperties.Add(autoProperty);
+                        }
+                    }
+                }
+
+                if (numFieldsRequiringInitialization == 0)
                 {
-                    // We have less field assignments than actual fields.  Generate a call to the
-                    // default constructor as well.
-                    return ImmutableArray<SyntaxNode>.Empty;
+                    return false;
                 }
+
+                foreach (var initialized in parameterToExistingFieldMap.Values)
+                {
+                    if (initialized.IsImplicitlyDeclared || initialized.IsStatic)
+                    {
+                        continue;
+                    }
+
+                    switch (initialized)
+                    {
+                        case IFieldSymbol _:
+                        case IPropertySymbol property when autoProperties.Contains(property):
+                            numFieldsRequiringInitialization--;
+                            if (numFieldsRequiringInitialization == 0)
+                            {
+                                return false;
+                            }
+                            break;
+                    }
+                }
```

Comment:
Honestly, a simpler version might be to just collect the fields/auto-props that need to be initialized into a set, then just remove frm that set anything in parameterToExistingFieldMap.Values.  If you end up with nothing, you don't need hte consturctor.

In other words, this could all be written with a little linq AFAICT...  Basically:

```
var needsConstructor =containingTypeOpt.GetMembers().OfType<IFieldSymbol>().Where(f => !f.IsStatic).Select(f => f.AssociatedProperty ?? (ISymbol)f).Except(parameterToExistingFieldMap.Values).Any();
```

Reply:
I was thinking that set operations would be slower than a list, since I rarely have a struct with ten or more fields.

Reply:
Both the loop and the Linq query will likely take microseconds tops.  But i think the linq approach is just clearer in conveying what is going on.